core/src/main/java/org/elasticsearch/search/aggregations/bucket/DeferringBucketAggregator.java

```diff
@@ -0,0 +1,113 @@
+
+    protected DeferringBucketAggregator(String name, AggregatorFactories factories, SearchContext context, Aggregator parent,
+            List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) throws IOException {
+        super(name, factories, context, parent, pipelineAggregators, metaData);
+    }
+
+    @Override
+    protected void doPreCollection() throws IOException {
+        List<BucketCollector> collectors = new ArrayList<>();
+        List<BucketCollector> deferredCollectors = new ArrayList<>();
+        for (int i = 0; i < subAggregators.length; ++i) {
+            if (shouldDefer(subAggregators[i])) {
+                if (recordingWrapper == null) {
+                    recordingWrapper = getDeferringCollector();
+                }
+                deferredCollectors.add(subAggregators[i]);
+                subAggregators[i] = recordingWrapper.wrap(subAggregators[i]);
+            } else {
+                collectors.add(subAggregators[i]);
+            }
+        }
+        if (recordingWrapper != null) {
+            recordingWrapper.setDeferredCollector(deferredCollectors);
+            collectors.add(recordingWrapper);
+        }
+        collectableSubAggregators = BucketCollector.wrap(collectors);
+    }
+
+    public DeferringBucketCollector getDeferringCollector() {
+        // Default impl is a collector that selects the best buckets
+        // but an alternative defer policy may be based on best docs.
+        return new BestBucketsDeferringCollector(context());
+    }
+
+    /**
+     * This method should be overridden by subclasses that want to defer
+     * calculation of a child aggregation until a first pass is complete and a
+     * set of buckets has been pruned. Deferring collection will require the
+     * recording of all doc/bucketIds from the first pass and then the sub class
+     * should call {@link #runDeferredCollections(long...)} for the selected set
+     * of buckets that survive the pruning.
+     * 
+     * @param aggregator
+     *            the child aggregator
+     * @return true if the aggregator should be deferred until a first pass at
+     *         collection has completed
+     */
+    protected boolean shouldDefer(Aggregator aggregator) {
+        return false;
+    }
+
+    protected final void runDeferredCollections(long... bucketOrds) throws IOException {
+        // Being lenient here - ignore calls where there are no deferred
+        // collections to playback
+        if (recordingWrapper != null) {
+            recordingWrapper.replay(bucketOrds);
+        }
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        return null;
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long bucket) throws IOException {
+        return null;
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return null;
+    }
```

Comment:
can these methods either be abstract if they need to be implemented or throw an exception in their default impl so that it is easier to figure out they have been called by mistake?

Reply:
This was my mistake, I forgot to delete the methods that Eclipse auto-generated. These methods already declared by `AggregatorBase` and are required to be implemented by subclasses so they can just be removed here